File: python_code/mesh.py
# -*- coding: utf-8 -*-
"""
Created on Wed Jun 13 10:06:23 2018

@author: Anthony

mesh.py

create meshes for pipeline processing
running using demo.py for conveniency
"""
import math
import numpy as np
import scipy.io
from scipy.spatial import Delaunay

import reconstruct as rc


def mesh(GRAB,SCANDIR,UNDISTORTED,THRESH,XRANGE,TRITHRESH,NBR_PASS):
    ''' generate mesh and save as .mat for smoothing and MeshLab '''
    
    ''' 2. Unpack box limit for meshing part 1 '''
    x_up_limit = XRANGE[0]  #250
    x_low_limit = XRANGE[1]    #-250
    y_up_limit = XRANGE[2]     #250
    y_low_limit = XRANGE[3]    #-250
    z_up_limit = XRANGE[4]      #600
    z_low_limit = XRANGE[5]    #0
    
    
    # run reconstruction script
    print("\nReconstruction started...")
    X, xL, xR, L_color, R_color = rc.reconstruct(SCANDIR, THRESH)
    print("Reconstruction done. Meshing started..")
    
    
    ####################################################
    #
    # step 1: remove points outside known bounding box
    #
    npts = np.shape(X)[1]
    goodidx = []
            
    for i in range(npts):
        x_ok = X[0,i] < x_up_limit and X[0,i] > x_low_limit
        y_ok = X[1,i] < y_up_limit and X[1,i] > y_low_limit
        z_ok = X[2,i] < z_up_limit and X[2,i] > z_low_limit
        
        if x_ok and y_ok and z_ok:
            goodidx.append(i)
    
    # check if no index found
    assert(len(goodidx) >= 3), "Error: no points found within range, consider expand boudning box"
    
    # drop the bad points from both 2D and 3D list
    X = X[:,goodidx]
    xL = xL[:,goodidx]
    xR = xR[:,goodidx]
    
    # optional: switch xR with xL
    tri = Delaunay(xR.transpose())
    tridx = tri.simplices
    print('..', end='')
    
    ####################################################
    #
    # step 2: remove triangles with edges longer than threshold
    #
    goodtri = []
    k = 0
    
    for i in range(np.shape(tridx)[0]):
        # get vertex coordinates
        A = X[:,tridx[i,0]]
        B = X[:,tridx[i,1]]
        C = X[:,tridx[i,2]]
        
        # compute edge lengths
        AB = math.sqrt((A[0]-B[0])**2 + (A[1]-B[1])**2 + (A[2]-B[2])**2)
        BC = math.sqrt((B[0]-C[0])**2 + (B[1]-C[1])**2 + (B[2]-C[2])**2)
        AC = math.sqrt((A[0]-C[0])**2 + (A[1]-C[1])**2 + (A[2]-C[2])**2)
        
        if max([AB,BC,AC]) < TRITHRESH:
            k = k+1
            goodtri.append(i)
    
    print('..', end='')
    # shrink goodtri and remove bad trianagles
    goodtri = goodtri[:k]
    tridx = tridx[goodtri,:]
    
    
    # remove unreferenced points not apperaing in any triangles
    ptind = np.unique(np.reshape(tridx, (1,-1)))
    
    # remap indices to the new range
    X = X[:,ptind]
    xR = xR[:,ptind]
    xL = xL[:,ptind]
    
    # remap triangle vertex indices with 2D for loop
    idxmap = dict()
    for i,j in enumerate(ptind):
        idxmap[j] = i
    
    for i in range(np.shape(tridx)[0]):
        for j in range(np.shape(tridx)[1]):
            tridx[i,j] = idxmap[tridx[i,j]]
        
        if (i % 20000) == 0:
            print('..', end='')
    
    print("\nMeshing done.")
    
    
    ###########################################################
    
    # colorize result
    # CAUTION: cv2.imread use BGR color space
    #           MATLAB uses RGB
    print("Computing colors.")
    
    colorCount = np.shape(X)[1]
    xColor = np.zeros((3,colorCount))
    
    for i in range(colorCount):
        rx, ry = xR[:,i]
        lx, ly = xL[:,i]
        xColor[2,i] = 0.5*(R_color[ry,rx,0] + L_color[ly,lx,0]) #B
        xColor[1,i] = 0.5*(R_color[ry,rx,1] + L_color[ly,lx,1]) #G
        xColor[0,i] = 0.5*(R_color[ry,rx,2] + L_color[ly,lx,2]) #R
    
    
    # inverting normal for MeshLab
    print("Inverting surface normals for MeshLab.")
    trimsh = np.zeros(np.shape(tridx))
    
    for i in range(np.shape(tridx)[0]):
        trimsh[i,0] = tridx[i,0]
        trimsh[i,1] = tridx[i,2]
        trimsh[i,2] = tridx[i,1]
        
        
    ############################################################
    # save to MATLAB file for 3D viewing, smoothing, and mesh_2_ply
    print("Done! Use meshprocessing.m with MATLAB to view, process, and convert mesh results")
    scipy.io.savemat('../cache/mesh.mat', mdict={'X':X,"xColor":xColor,"tri":tridx.astype(float),"trimsh":trimsh.astype(float),"NBR_PASS":float(NBR_PASS),"SCANDIR":SCANDIR,"GRAB":GRAB,})
    print("Grab number %d" % GRAB)


    return 0

File: matlab_code/mesh.m
function [] = mesh(GRAB,SCANDIR,UNDISTORTED,THRESH,XRANGE,TRITHRESH,NBR_PASS)
    % mesh.m - compute meshing with triangulated points
    % Input: parameters from demo.m
    % Output: None, directly to .ply files
    % 

    % Unpack bound box limit for part 1
    x_max = XRANGE(1);  %250
    x_min = XRANGE(2);  %-250
    y_max = XRANGE(3); %250
    y_min = XRANGE(4); %-250
    z_max = XRANGE(5); %600
    z_min = XRANGE(6); %0

    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    % reconstruct surface
    fprintf('Reconstruction started...\n');
    reconstruct(UNDISTORTED, SCANDIR, THRESH);
    load reconstruction.mat
    fprintf("Reconstruction done\n");

    % view reconstructed points
    clf; plot3(X(1,:),X(2,:),X(3,:),'.');
    axis image; axis vis3d; grid on;
    hold on;
    plot3(camL.t(1),camL.t(2),camL.t(3),'go')
    plot3(camR.t(1),camR.t(2),camR.t(3),'go')

    % plot bounding box vertices
    plot3(x_min,y_min,z_min,'r*')
    plot3(x_min,y_min,z_max,'r*')
    plot3(x_min,y_max,z_min,'r*')
    plot3(x_min,y_max,z_max,'r*')
    plot3(x_max,y_min,z_min,'r*')
    plot3(x_max,y_min,z_max,'r*')
    plot3(x_max,y_max,z_min,'r*')
    plot3(x_max,y_max,z_max,'r*')

    axis([x_min-200 200+x_max y_min-200 200+y_max z_min-200 200+z_max])
    set(gca,'projection','perspective')
    xlabel('X-axis');
    ylabel('Y-axis');
    zlabel('Z-axis');

    fprintf(" verify the bounding box vertices\n press 'Enter' to continue..\n")
    pause;   
    fprintf("Bounding box verified, continuing.\n")


    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    %
    % cleaning step 1: remove points outside known bounding box
    %
    fprintf("Meshing started.\n")

    goodind = find((X(1,:)<x_max) & (X(1,:)>x_min) & (X(2,:)<y_max) & (X(2,:)>y_min) & (X(3,:)<z_max) & (X(3,:)>z_min));

    % check if no points found
    assert(size(goodind,2) >= 3, "Error: no points found within range. Consider expanding bounding box.");

    %
    % drop bad points from both 2D and 3D list
    %

    X = X(:,goodind);
    xL = xL(:,goodind);
    xR = xR(:,goodind);

    tri = delaunay(xR(1,:), xR(2,:));

    fprintf('.');

    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    %
    % cleaning step 2: remove triangles which have long edges
    %
    goodtri = zeros(1,size(tri,1));
    k = 0;

    for i = 1:size(tri,1)
        % acquire point coordinates
        A = X(:,tri(i,1));
        B = X(:,tri(i,2));
        C = X(:,tri(i,3));

        % compute edge lengths
        AB = sqrt( (A(1)-B(1))^2 + (A(2)-B(2))^2 + (A(3)-B(3))^2 );
        BC = sqrt( (B(1)-C(1))^2 + (B(2)-C(2))^2 + (B(3)-C(3))^2 );
        CA = sqrt( (C(1)-A(1))^2 + (C(2)-A(2))^2 + (C(3)-A(3))^2 );

        if max([AB, BC, CA]) < TRITHRESH
            k = k + 1;
            goodtri(k) = i;
        end
    end
    fprintf('.');
    % shrink goodtri and remove bad triangles
    goodtri = goodtri(1:k);
    tri = tri(goodtri,:); 


    % remove unreferenced points which don't appear in any triangle

    ptind = unique(reshape(tri,1,[])); % retrieve points of triangles

    % remap indices to new range
    X = X(:,ptind); 
    xR = xR(:,ptind);
    xL = xL(:,ptind);

    % remap triangle vertex indices with 2D for loop
    fprintf('.');

    for i = 1:size(tri,1)
        for j = 1:size(tri,2)
            tri(i,j) = find(ptind==tri(i,j));
        end

        if rem(i, 20000) == 0
            fprintf('.');
        end
    end

    fprintf('\nMeshing done.\n');

    fprintf('nbr_smooth\n');
    X = nbr_smooth(tri, X, NBR_PASS);
    fprintf('nbr_smooth done.\n');

    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    %
    % display results
    %

    figure(1); clf;
    h = trisurf(tri,X(1,:),X(2,:),X(3,:));
    set(h,'edgecolor','none');
    set(gca,'projection','perspective');
    xlabel('X-axis');
    ylabel('Y-axis');
    zlabel('Z-axis');
    axis image; axis vis3d;


    % rotate the view around so we see from
    % the front  (can also do this with the mouse in the gui)
    camorbit(45,0);
    camorbit(0,-120);
    camroll(-8);

    % prepare output savings
    fprintf('Saving mesh...\n');
    colorCount = size(xR, 2);
    xColor = zeros(3,colorCount);

    for i = 1:colorCount
        xr = xR(1,i);
        yr = xR(2,i);
        xl = xL(1,i);
        yl = xL(2,i);
        xColor(1,i) = 0.5*(R_color(yr,xr,1) + L_color(yl,xl,1));
        xColor(2,i) = 0.5*(R_color(yr,xr,2) + L_color(yl,xl,2));
        xColor(3,i) = 0.5*(R_color(yr,xr,3) + L_color(yl,xl,3));
    end

    % this is to invert the surface normals for MeshLab
    tri_MeshLab = zeros(size(tri));

    for i = 1:size(tri,1)
        tri_MeshLab(i,1) = tri(i,1);
        tri_MeshLab(i,2) = tri(i,3);
        tri_MeshLab(i,3) = tri(i,2);
    end

    % call mesh_2_ply.m
    mesh_2_ply( X,xColor,tri_MeshLab,[SCANDIR sprintf('mesh%d.ply',GRAB)])
    fprintf('Mesh successfully saved. DONE!\n');

end

